### PR TITLE
Preserve data in the browser date 12/23/2022

### PR DESCRIPTION
**Pull request by fithamlak**
=>Implement all requirements for Preserve data in the browser project and all passed like
  -  get form elements and check local storage
  - create a single object for contact data and listen to change on input fields
  - finally stringify that data and set the object to local on loads of page
coding-partner-challenge

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -41,4 +41,92 @@ form.addEventListener('submit', (event) => {
...
+function retrieveData() {
+  const contactData = availableStorage.getItem('FormData');
+  const parseContactData = JSON.parse(contactData);
+  if (contactData?.length > 0) {
+    const { name, email, message } = parseContactData;
+    nameInput.value = name || '';
+    emailInput.value = email || '';
+    messageInput.value = message || '';
+  }
+}
```
Comment (Tadesse-Alemayehu) [inline index.js:128]: Dear @fithamlak, Good job so far, As part of the main requirement for this project we should be able to store and retrieve data from local storage. Your form is saving the data only once and only if hit submit. on the next reload the content are not available. Please address this for a better user experience. 
### Remember the following point as you work on the contact form. 

- [ ] The contact form information is persistent to page reloads, This includes after filling the form and not submitting it.
- [ ] When visiting the page the form inputs should be contained, be auto-populated with, the values entered in the previous visit to the page. and If at the previous visit to the page, in the same browser, no input was entered in the form then the form inputs should remain empty.
- [ ] To achieve both features that I have mentioned you could make use of `localStorage.getItem` and `localStorage.setItem`

**Review by Tadesse-Alemayehu — CHANGES_REQUESTED** [review]
### Hi @fithamlak 
### I'm Tadesse Alemayehu your reviewer for this PR. :hand: 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!
EquatorialVictoriousBlackpanther-size_restricted

## Highlights
 - [ ]  Correct GitHub Flow  ✅
 - [ ]  Good commit history  ✅
 - [ ]  Good PR title/description  ✅ 
## Required Changes ♻️
- [ ] Dear @fithamlak, As part of this milestone we need to have the JavaScript linter `Eslint` Please add the JavaScript linter to your project.

_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._
Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: index.js [hunk 2]
```diff
@@ -41,4 +41,92 @@ form.addEventListener('submit', (event) => {
...
+  storeData();
+});
+// retrive the data by stringify and
+// set the object to local on loads of page
+function retrieveData() {
+  const contactData = availableStorage.getItem('FormData');
+  const parseContactData = JSON.parse(contactData);
+  if (contactData?.length > 0) {
+    const { name, email, message } = parseContactData;
+    nameInput.value = name || '';
```
Comment (Olamarx) [inline index.js:124]: - [ ] As the previous reviewer observed, the data is not getting and I also found out that it is also not saved in the localStorage. I tried to check your code further, but I could not find where you declared `nameInput` I see that you added eventLister to it on line 106 and my console has an error with it.
Kindly dig further into the tag you are adding to the event listener.
### The console error
image

**Review by Olamarx — CHANGES_REQUESTED** [review]
Hi @fithamlak, 

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.


 ## Highlights
- Linters passed ✔️
- Readme file ✔️

 ## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by KayLemba — APPROVED** [review]
[# STATUS: APPROVED :green_circle: :clap: 

Hello  @fithamlak        , 
Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

HIGHLIGHTS:

- [x]  Your repo is professional 
- [x] Descriptive P.R
- [x] Descriptive README file




 








Now you can merge this branch and move on :1st_place_medal: 
Happy Coding cheers 🥇 💯 



well-1

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._]

**Comment by huseyinbiyik** [comment]
Hi @fithamlak,

While you made a great effort in this project, unfortunately, I cannot proceed to review your code.

Alt Text

## Invalid Code Review Request

❗️ You have submitted a project, that has a missing section in your Readme. Please make sure you added the section Future Features, including 1-3 potential features to be added. Please notice that even though it is approved I need to check everything like it is the first review. 


Your Code Review Request will be marked as invalid in your Dashboard, so please submit a new one once you are ready 🙏

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

--------

_Invalid Code Review Request **does not count** into the code reviews limit._

**Pull request merged**